File: Backend/src/main/java/com/example/Backend/service/ProductService.java
package com.example.Backend.service;

import com.example.Backend.dto.ProductDTO;
import com.example.Backend.entity.Product;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

public interface ProductService {

    /**
     * Create a new product
     * 
     * @param productDTO Product data
     * @return Created product
     */
    Product createProduct(ProductDTO productDTO);

    /**
     * Get product by ID
     * 
     * @param id Product ID
     * @return Product entity
     */
    Product getProductById(UUID id);

    /**
     * Update product
     * 
     * @param id         Product ID
     * @param productDTO Updated product data
     * @return Updated product
     */
    Product updateProduct(UUID id, ProductDTO productDTO);

    /**
     * Delete product
     * 
     * @param id Product ID
     */
    void deleteProduct(UUID id);

    /**
     * Get all products
     * 
     * @return List of all products
     */
    List<Product> getAllProducts();

    /**
     * Get products with pagination
     * 
     * @param pageable Pagination parameters
     * @return Page of products
     */
    Page<Product> getAllProducts(Pageable pageable);

    /**
     * Search products with multiple filters
     * 
     * @param searchTerm Search term (name/description)
     * @param category   Category filter
     * @param brand      Brand filter
     * @param model      Model filter
     * @param minPrice   Minimum price
     * @param maxPrice   Maximum price
     * @param pageable   Pagination parameters
     * @return Page of matching products
     */
    Page<Product> searchProducts(String searchTerm, String category, String brand,
            String model, BigDecimal minPrice, BigDecimal maxPrice,
            Pageable pageable);

    /**
     * Filter products by category, brand, price range, stock, and search term
     * 
     * @param categoryId Category ID
     * @param brandId    Brand ID
     * @param minPrice   Minimum price
     * @param maxPrice   Maximum price
     * @param search     Search term
     * @param inStock    Filter by stock availability
     * @param pageable   Pagination parameters
     * @return Page of filtered products
     */
    Page<Product> filterProducts(Long categoryId, Long brandId, BigDecimal minPrice,
            BigDecimal maxPrice, String search, Boolean inStock, Pageable pageable);

    /**
     * Get products by category
     * 
     * @param category Category name
     * @return List of products
     */
    List<Product> getProductsByCategory(String category);

    /**
     * Get products by brand
     * 
     * @param brand Brand name
     * @return List of products
     */
    List<Product> getProductsByBrand(String brand);

    /**
     * Get products compatible with vehicle
     * 
     * @param brand Vehicle brand
     * @param model Vehicle model
     * @param year  Vehicle year
     * @return List of compatible products
     */
    List<Product> getCompatibleProducts(String brand, String model, Integer year);

    /**
     * Get products in stock
     * 
     * @return List of products with stock > 0
     */
    List<Product> getProductsInStock();

    /**
     * Get low stock products
     * 
     * @param threshold Stock threshold
     * @return List of low stock products
     */
    List<Product> getLowStockProducts(Integer threshold);

    /**
     * Get out of stock products
     * 
     * @return List of out of stock products
     */
    List<Product> getOutOfStockProducts();

    /**
     * Get top selling products
     * 
     * @param pageable Pagination parameters
     * @return Page of top selling products
     */
    Page<Product> getTopSellingProducts(Pageable pageable);

    /**
     * Get featured products
     * 
     * @param pageable Pagination parameters
     * @return Page of featured products
     */
    Page<Product> getFeaturedProducts(Pageable pageable);

    /**
     * Get all categories
     * 
     * @return List of distinct categories
     */
    List<String> getAllCategories();

    /**
     * Get all brands
     * 
     * @return List of distinct brands
     */
    List<String> getAllBrands();

    /**
     * Get models for brand
     * 
     * @param brand Brand name
     * @return List of models
     */
    List<String> getModelsByBrand(String brand);

    /**
     * Update product stock
     * 
     * @param id       Product ID
     * @param quantity Quantity to add (positive) or subtract (negative)
     * @return Updated product
     */
    Product updateStock(UUID id, Integer quantity);

    /**
     * Count products in stock
     * 
     * @return Number of products in stock
     */
    Long countProductsInStock();
}

File: frontend-web/src/app/core/services/product.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { Product, ProductResponse, ProductFilter, Category, Brand } from '../models';

/**
 * Product Service
 * Handles all product-related API calls
 */
@Injectable({
    providedIn: 'root'
})
export class ProductService {
    private readonly endpoint = 'products';
    private readonly categoryEndpoint = 'categories';
    private readonly brandEndpoint = 'brands';

    constructor(private apiService: ApiService) { }

    /**
     * Get all products with optional filters and pagination
     */
    getProducts(filter?: ProductFilter): Observable<ProductResponse> {
        const params: any = {};

        if (filter) {
            if (filter.page !== undefined) params.page = filter.page.toString();
            if (filter.size !== undefined) params.size = filter.size.toString();
            if (filter.search) params.search = filter.search;
            if (filter.categoryId) params.categoryId = filter.categoryId.toString();
            if (filter.brandId) params.brandId = filter.brandId.toString();
            if (filter.minPrice !== undefined) params.minPrice = filter.minPrice.toString();
            if (filter.maxPrice !== undefined) params.maxPrice = filter.maxPrice.toString();
            if (filter.inStock !== undefined) params.inStock = filter.inStock.toString();
            if (filter.sort) params.sort = filter.sort;
        }

        return this.apiService.get<ProductResponse>(this.endpoint, params);
    }

    /**
     * Get a single product by ID
     */
    getProductById(id: number | string): Observable<Product> {
        // If id is a number, it's being used as-is (may fail if backend expects UUID)
        // If id is a string (UUID), use it directly
        return this.apiService.get<Product>(`${this.endpoint}/${id}`);
    }

    /**
     * Get featured products
     */
    getFeaturedProducts(limit: number = 8): Observable<Product[]> {
        return this.apiService.get<Product[]>(`${this.endpoint}/featured`, { limit: limit.toString() });
    }

    /**
     * Search products by query
     */
    searchProducts(query: string, page: number = 0, size: number = 20): Observable<ProductResponse> {
        return this.apiService.get<ProductResponse>(`${this.endpoint}/search`, {
            q: query,
            page: page.toString(),
            size: size.toString()
        });
    }

    /**
     * Get products by category
     */
    getProductsByCategory(categoryId: number, page: number = 0, size: number = 20): Observable<ProductResponse> {
        return this.apiService.get<ProductResponse>(`${this.endpoint}/category/${categoryId}`, {
            page: page.toString(),
            size: size.toString()
        });
    }

    /**
     * Get products by brand
     */
    getProductsByBrand(brandId: number, page: number = 0, size: number = 20): Observable<ProductResponse> {
        return this.apiService.get<ProductResponse>(`${this.endpoint}/brand/${brandId}`, {
            page: page.toString(),
            size: size.toString()
        });
    }

    /**
     * Get all categories
     */
    getCategories(): Observable<Category[]> {
        return this.apiService.get<Category[]>(this.categoryEndpoint);
    }

    /**
     * Get all brands
     */
    getBrands(): Observable<Brand[]> {
        return this.apiService.get<Brand[]>(this.brandEndpoint);
    }

    /**
     * Get related products
     */
    getRelatedProducts(productId: number, limit: number = 4): Observable<Product[]> {
        return this.apiService.get<Product[]>(`${this.endpoint}/${productId}/related`, {
            limit: limit.toString()
        });
    }
}
